controllers/azuremachine_reconciler.go

```diff
@@ -78,36 +78,26 @@ func (s *azureMachineService) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "controllers.azureMachineService.Reconcile")
 	defer span.End()
 
-	if err := s.publicIPsSvc.Reconcile(ctx); err != nil {
-		return errors.Wrap(err, "failed to create public IP")
+	ar := newAsyncReconciler()
+
+	ar.submit(ctx, s.publicIPsSvc)
+	ar.submit(ctx, s.inboundNatRulesSvc)
+	ar.submit(ctx, s.networkInterfacesSvc)
+	ar.submit(ctx, s.availabilitySetsSvc)
+	if err := ar.wait(); err != nil {
+		return err
 	}
 
```

Comment:
what about dependencies between these resources

Reply:
The resources that are independent are submitted to the asyncReconciler which will be processed in parallel, and whenever there's a dependency, we wait for the tasks to finish before proceeding.